Update json-spirit to latest version to support multibyte characters.

rkk-ableton opened the pull request:
## Summary

Updates json-spirit to the latest public version and adds tests to validate the change.

## Details

This change updates json-spirit to the latest public version:
https://www.codeproject.com/KB/recipes/JSON_Spirit/json_spirit_v4.08.zip

In addition to the updated dependency, the call to write_string inside
WireProtocol::encode is modified to pass the raw_utf8 option by default.

There are also two new tests added which validate the functionality of passing UTF-8 encoded multibyte character strings over the WireProtocol.

## Motivation and Context

Previously, multibyte characters were being escaped before transmission
between the cucumber rubygem and cucumber-cpp. This broke test
definitions which relied on multibyte characters being
transmitted through the WireProtocol unescaped.

Commit:
Update json-spirit to version 4.08 to support multibyte characters.

This change updates json-spirit to the latest public version:
https://www.codeproject.com/KB/recipes/JSON_Spirit/json_spirit_v4.08.zip

4.08 adds support for a raw_utf8 option when writing a JSON string.
Previously, multibyte characters were being escaped when being sent from
cucumber-cpp to cucumber-ruby. Because cucumber-ruby's wire decoder
does not properly decode escaped character sequences, this would crash
cucumber-ruby.

ala-ableton commented:
@src-ableton `json_spirit_writer_options.h` is in `include/json_spirit/` instead of being in `3rdparty/json_spirit/`, which means you moved files manually instead of extracting them from the mentioned archive in the right place. Please make this commit again.

Commit:
WireResponseEncoder::encode now outputs raw utf8 by default.

This modifies the WireResponseEncoder to always use the raw_utf8
option provided by the new version of json_spirit.

According to the IETF RFC8259:
"JSON text exchanged between systems that are not part of a closed
   ecosystem MUST be encoded using UTF-8 [RFC3629]."
https://tools.ietf.org/html/rfc8259

Commit:
Recalculate position so it is based on the index of unicode codepoints.

`cucumber-ruby` expects position values which are based on the index of
the codepoint instead of the index of the code unit. This change modifies
the value returned to `cucumber-ruby`.

Prior to this change, the RegexSubMatch's position, which was correct in
terms of a code unit array, would cause an `index out of string` error and
crash cucumber-ruby when pretty-printing the results of a test.

This commit also ammends the added tests to demonstrate the corrected
behavior.

rkk-ableton commented:
@ala-ableton @mtt-ableton @dsr-ableton @nil-ableton @miv-ableton RFC

miv-ableton reviewed (commented):
LGTM, but I'm definitely not an expert of this code.

dsr-ableton commented on src/StepManager.cpp:
```diff
@@ -114,12 +115,34 @@ MatchResult StepManager::stepMatches(const std::string &stepDescription) {
+
+std::ptrdiff_t StepManager::codePointsFromStart(const std::string &expression, size_t index) {
+    std::ptrdiff_t codePointCount(0);
+    for (const uint8_t i : expression) {
```
Traditionally, chars were named `c` instead of `i`. 🐦

ala-ableton replied:
@src-ableton `cucumber-cpp` still supports C++98, you can't use for-range loops.

dsr-ableton commented on include/cucumber-cpp/internal/step/StepManager.hpp:
```diff
@@ -183,6 +183,7 @@ class CUCUMBER_CPP_EXPORT StepManager {
 public:
     static step_id_type addStep(boost::shared_ptr<StepInfo> stepInfo);
     static MatchResult stepMatches(const std::string &stepDescription);
+    static std::ptrdiff_t codePointsFromStart(const std::string &expression, size_t index);
```
commit msg: To me "codepoint position" or "codepoint offset" would sound more natural.

ala-ableton commented on src/StepManager.cpp:
```diff
@@ -114,12 +115,34 @@ MatchResult StepManager::stepMatches(const std::string &stepDescription) {
         const boost::shared_ptr<const StepInfo>& stepInfo = iter->second;
         SingleStepMatch currentMatch = stepInfo->matches(stepDescription);
         if (currentMatch) {
+            for (auto& submatch : currentMatch.submatches) {
```
@src-ableton `cucumber-cpp` still supports C++98, you can't use for-range loops.

nil-ableton replied:
cf https://travis-ci.org/AbletonAppDev/cucumber-cpp/jobs/545210181

rkk-ableton replied:
No longer using any syntax from C++ >0x.

nil-ableton commented:
commit message for e18e6028278b838c50eba6ed332c60798a5a0c01 mentions using raw_utf8 which I think belongs to later commits.

nil-ableton commented:
Reading https://github.com/AbletonAppDev/cucumber-cpp/pull/8/commits/7d01dc5b7b33c55ad4f88c3d12608a86be8f2473 leaves me with some question.

We are claiming here that we must use raw_utf8 format because that's what `cucumber-ruby` is expecting. I assume this also means that `cucumber-ruby` does NOT support the escaped multibyte
characters (`\uXYZW`) that json spirit claims to emit when `raw_utf8` is not used. Am I correct?

Since I wanted to check how and why raw_utf8 was required, I went to check what cucumber-ruby was saying. First, I think I understood that the wire protocol is implemented in `cucumber-ruby-wire` instead.

The following file shows no sign of forcing any sort of character format:
  https://github.com/cucumber/cucumber-ruby-wire/blob/master/lib/cucumber/wire/data_packet.rb

Additionally that project looks like it's using a `multi_json` ruby library https://github.com/intridea/multi_json ; this creates other questions: that library claims to be generic and support various kinds of json protocols (I assume dynamically, depending on what the user has installed)

So this makes me suspicious: is it really a static property of the cucumber-ruby project that they would work with raw_utf8? Is it possible that raw_utf8 would be harmful in some other situations (for instance when using a different underlying json parser discovered by multi_json, be it because we used a different ruby etc...)

So 1) I'm not sure we're being precise enough about why we need `raw_utf8` in our commit (is it because of multi_json's underlying json library that is used in our particular setup? 2) I'm not sure whether `raw_utf8` will be acceptable for the `cucumber-cpp` if for some reason it would not be accepted by other users' underlying json parsers.

nil-ableton commented on src/StepManager.cpp:
```diff
@@ -114,12 +115,34 @@ MatchResult StepManager::stepMatches(const std::string &stepDescription) {
+        if (i < 0x80) {
+            codePointCount++;
+        }
+        if ((i & 0b1100'0000) == 0b1100'0000) {
```
https://en.cppreference.com/w/cpp/language/integer_literal

- 0b (binary literals) are >= c++14
- digit separators also are

So you'll have to convert it to an hexadecimal litteral, unfortunately

rkk-ableton commented:
> Reading [7d01dc5](https://github.com/AbletonAppDev/cucumber-cpp/commit/7d01dc5b7b33c55ad4f88c3d12608a86be8f2473) leaves me with some question.
> 
> We are claiming here that we must use raw_utf8 format because that's what `cucumber-ruby` is expecting. I assume this also means that `cucumber-ruby` does NOT support the escaped multibyte
> characters (`\uXYZW`) that json spirit claims to emit when `raw_utf8` is not used. Am I correct?
> 
> Since I wanted to check how and why raw_utf8 was required, I went to check what cucumber-ruby was saying. First, I think I understood that the wire protocol is implemented in `cucumber-ruby-wire` instead.
> 
> The following file shows no sign of forcing any sort of character format:
> https://github.com/cucumber/cucumber-ruby-wire/blob/master/lib/cucumber/wire/data_packet.rb
> 
> Additionally that project looks like it's using a `multi_json` ruby library https://github.com/intridea/multi_json ; this creates other questions: that library claims to be generic and support various kinds of json protocols (I assume dynamically, depending on what the user has installed)
> 
> So this makes me suspicious: is it really a static property of the cucumber-ruby project that they would work with raw_utf8? Is it possible that raw_utf8 would be harmful in some other situations (for instance when using a different underlying json parser discovered by multi_json, be it because we used a different ruby etc...)
> 
> So 1) I'm not sure we're being precise enough about why we need `raw_utf8` in our commit (is it because of multi_json's underlying json library that is used in our particular setup? 2) I'm not sure whether `raw_utf8` will be acceptable for the `cucumber-cpp` if for some reason it would not be accepted by other users' underlying json parsers.

@nil-ableton I understand your concerns and agree about the potential risks to other users who may have different JSON parsers. I'm curious what you think of moving the encoder options to be an argument to the constructor of the WireProtocol?

If you think it's worthwhile, I'll implement/test it and update the PR with the change.

nil-ableton commented:
> moving the encoder options to be an argument to the constructor of the WireProtocol?

@src-ableton that solution is appealing because it would allow us to use it in our safe/static environment. I find it convincing that it would help landing the change

rkk-ableton commented:
> that solution is appealing because it would allow us to use it in our safe/static environment. I find it convincing that it would help landing the change

@nil-ableton Looking into this further, I think defaulting to raw_utf8 is the correct approach. The JSON spec requires that parsers support UTF-8, so transmitting multibyte characters without escaping them is valid JSON behavior.

nil-ableton commented:
Yeah after reading the rfc8259 it looks like a reasonable position to take

rkk-ableton commented:
@toh-ableton RFC when you have some time.

toh-ableton commented on 3rdparty/json_spirit/CMakeLists.txt:
```diff
@@ -8,6 +8,7 @@ target_sources(json_spirit.header INTERFACE
     ${CMAKE_CURRENT_SOURCE_DIR}/json_spirit_stream_reader.h
     ${CMAKE_CURRENT_SOURCE_DIR}/json_spirit_utils.h
     ${CMAKE_CURRENT_SOURCE_DIR}/json_spirit_value.h
+    ${CMAKE_CURRENT_SOURCE_DIR}/json_spirit_writer_options.h
```
Just a nit on the commit message:
> This broke test definitions which relied on multibyte characters being **accurately** transmitted through the WireProtocol.

I don't think `accurate` is the correct term. JSON explicitly allows escaped characters, so the problem is rather that cucumber-ruby doesn't correctly deal with escaped characters if I understand the issue correctly.

toh-ableton commented on src/connectors/wire/WireProtocol.cpp:
```diff
@@ -3,6 +3,7 @@
 
 #include <json_spirit/json_spirit_reader_template.h>
 #include <json_spirit/json_spirit_writer_template.h>
+#include <json_spirit/json_spirit_writer_options.h>
```
Again, I think the commit message is slightly off-point. It's not that the text wasn't UTF-8 encoded before, but rather that codepoints above U+007F were escaped. The reason for the change is that the cucumber-ruby wire decoder doesn't correctly decode escaped character sequences.

rkk-ableton replied:
The message has been updated to phrase better.

toh-ableton commented on src/connectors/wire/WireProtocol.cpp:
```diff
@@ -231,7 +232,7 @@ namespace {
             response.accept(*this);
             const mValue v(jsonOutput);
-            return write_string(v, false);
+            return write_string(v, json_spirit::Output_options::raw_utf8);
```
Please add a test for this change. Ideally, this would be a unit test (a new test case in `tests/WireProtocolTest.cpp`) plus an acceptance test (something like a new feature file `features/specific/wire-encoding.feature`).

toh-ableton commented on src/StepManager.cpp:
```diff
@@ -114,12 +116,36 @@ MatchResult StepManager::stepMatches(const std::string &stepDescription) {
 
+
+std::ptrdiff_t StepManager::codepointOffset(const std::string &expression, size_t byteIndex) {
+    assert(byteIndex < expression.size());
```
I don't see any other `asserts` in `cucumber-cpp`. I believe their error reporting policy is to `throw` a `std::range_error` in this situation.

toh-ableton commented on src/StepManager.cpp:
```diff
@@ -114,12 +116,36 @@ MatchResult StepManager::stepMatches(const std::string &stepDescription) {
+            codePointCount++;
+        }
+    }
+    return codePointCount;
```
Why not simply this instead:
```c++
#include <algorithm>

bool isStartOfCodepoint(char c) {
    return (c & 0xc0) != 0x80;
}

std::ptrdiff_t codepointIndex(const std::string &expression, size_t byteIndex) {
  if (byteIndex > expression.size()) {
    throw std::range_error("byteIndex out of range");
  }
  return count_if(expression.begin(), expression.begin() + byteIndex, &isStartOfCodepoint);
}
```
❓

rkk-ableton replied:
Thanks for the snippet. The code is updated with your suggestions (I made some modifications).

toh-ableton commented on include/cucumber-cpp/internal/step/StepManager.hpp:
```diff
@@ -183,6 +183,7 @@ class CUCUMBER_CPP_EXPORT StepManager {
 public:
     static step_id_type addStep(boost::shared_ptr<StepInfo> stepInfo);
     static MatchResult stepMatches(const std::string &stepDescription);
+    static std::ptrdiff_t codepointOffset(const std::string &expression, size_t byteIndex);
```
Two questions:
1. Why is this a member function?
2. Why is this function public? If the function needs to be public, please add a unit test.

rkk-ableton replied:
1. It doesn't need to be, but I didn't see any non-member functions defined in StepManager.cpp.
2. See 1.

It's no longer a member function.

toh-ableton commented on include/cucumber-cpp/internal/step/StepManager.hpp:
```diff
@@ -183,6 +183,7 @@ class CUCUMBER_CPP_EXPORT StepManager {
 public:
     static step_id_type addStep(boost::shared_ptr<StepInfo> stepInfo);
     static MatchResult stepMatches(const std::string &stepDescription);
```
Please add a test case that demonstrates the bug to `tests/unit/StepManagerTest.cpp`. Ideally, please also add a test case to your new `features/specific/wire-encoding.feature`.

rkk-ableton replied:
Both of these now exist (created on the [test-multibyte-char-strings](https://github.com/AbletonAppDev/cucumber-cpp/tree/test-multibyte-char-strings) branch so they can be tested in isolation).

toh-ableton reviewed (commented):
I only have a few requests before I think this change is ready:

1. Please add test coverage for this new feature. I indicated inline what I thought would be appropriate.
2. The implementation of StepManager::codepointOffset seems unnecessarily complicated. If it remains public, it needs to be tested.
3. I have a few nits about the commit messages.
4. Please take a look at https://github.com/cucumber/cucumber-cpp/issues/40 and figure out if or how much it fixes of this issue. This is going to be of interest to the cucumber-cpp maintainers. Please mention it when opening the PR upstream.

Overall, this is looking great and moving in a good direction!

Commit:
Add tests to demonstrate bug caused by json_spirit.

json_spirit's escaping of multibyte characters creates bugs in the WireProtocol
which prevent usage of valid UTF-8 encoded characters in step definitions
relying on RegEx.

The new tests:
/features/specific/wire_encoding.feature
/tests/integration/WireProtocolTest.cpp
/tests/unit/RegexTest.cpp

nil-ableton commented on src/StepManager.cpp:
```diff
@@ -108,12 +110,31 @@ step_id_type StepManager::addStep(boost::shared_ptr<StepInfo> stepInfo) {
+    if (byteIndex >= expression.size()) {
+        throw std::range_error("byteIndex out of range");
+    }
+    return count_if(expression.begin(), expression.begin() + byteIndex, &isStartOfCodepoint);
```
How come you did not need `std::` prefix before the count_if ?

nil-ableton replied:
Answer: this is because ADL is kicking in, when recognizing the iterators

rkk-ableton commented:
@ala-ableton Travis is failing because of some `clang-format` issues. Since this is a third-party library, does it make sense to disable that CI step?

ala-ableton commented:
@src-ableton I pushed a few fixup commits that should help.

dsr-ableton commented on src/Regex.cpp:
```diff
@@ -25,6 +25,19 @@ boost::shared_ptr<RegexMatch> Regex::find(const std::string &expression) const {
 }
 
+namespace {
+bool isStartOfCodepoint(unsigned int i) {
```
This is tailored to UTF-8, so please put that information somewhere, e.g. `utf8CodeUnit` instead of `i` or `isUtf8CodeUnitStartOfCodepoint`.

dsr-ableton replied:
Haha, in mtt-japanese-search-support, I found this even shorter test: `(i & 0xc0) != 0x80)` for the start of a code point.

toh-ableton replied:
https://github.com/AbletonAppDev/cucumber-cpp/pull/8#discussion_r293476620 😎

dsr-ableton commented on src/Regex.cpp:
```diff
@@ -25,6 +25,19 @@ boost::shared_ptr<RegexMatch> Regex::find(const std::string &expression) const {
+    return i < 0x80 || (i & 0xC0) == 0xC0;
+}
+
+std::ptrdiff_t codepointOffset(const std::string& expression, size_t bytePosition) {
```
Please standardise on `ptrdiff_t`. Especially because you cast from `std::ptrdiff_t` to `std::size_t` at the call site.

Also 🐦: This code assumes that `expression` is UTF-8 encoded and that characters are 8-bit values. Both of these hold, but I don't see why you talk about bytes here. Technically it's correct, but we know much more about these "bytes": They are UTF-8 code units. This is much more specific than "bytes". This also applies to the commit message.

dsr-ableton commented on src/Regex.cpp:
```diff
@@ -37,7 +50,8 @@ FindRegexMatch::FindRegexMatch(const boost::regex &regexImpl, const std::string
             if (i->matched) {
-                RegexSubmatch s = {*i, i->first - expression.begin()};
+                std::ptrdiff_t bytePosition = i->first - expression.begin();
+                RegexSubmatch s = {*i, codepointOffset(expression, bytePosition)};
```
Now that this constructor throws, please adjust its declaration, cf. `internal/Table.hpp`: `addColumn(·)`
Also 🐦: `bytePosition`, see above.

dsr-ableton commented on src/Regex.cpp:
```diff
@@ -25,6 +25,19 @@ boost::shared_ptr<RegexMatch> Regex::find(const std::string &expression) const {
+
+std::ptrdiff_t codepointOffset(const std::string& expression, size_t bytePosition) {
+    if (bytePosition > expression.size()) {
+        throw std::range_error("bytePosition out of range");
```
&nbsp;+ `#include <stdexcept>`

dsr-ableton commented on src/StepManager.cpp:
```diff
@@ -1,5 +1,7 @@
 #include "cucumber-cpp/internal/step/StepManager.hpp"
 
+#include <algorithm>
```
Why is this suddenly necessary?

dsr-ableton commented on src/connectors/wire/WireProtocol.cpp:
```diff
@@ -231,7 +232,7 @@ namespace {
             jsonOutput.clear();
             response.accept(*this);
             const mValue v(jsonOutput);
-            return write_string(v, false);
```
Shouldn't this commit also change at least `WireProtocolTest` to make it pass again?

dsr-ableton commented on src/Regex.cpp:
```diff
@@ -25,6 +25,19 @@ boost::shared_ptr<RegexMatch> Regex::find(const std::string &expression) const {
+    if (bytePosition > expression.size()) {
+        throw std::range_error("bytePosition out of range");
+    }
+    return count_if(expression.begin(), expression.begin() + bytePosition, &isStartOfCodepoint);
```
Also: Shouldn't this modify some tests to make them pass again?

ala-ableton committed:
fixup! Recalculate position so it is based on the index of unicode codepoints.

ala-ableton commented on travis.sh:
```diff
@@ -40,7 +40,7 @@ if [ -n "${FORMAT:-}" ]; then
     git clang-format-3.8 --binary=/usr/bin/clang-format-3.8 --style=file --commit="${BASE_HEAD}"
     # Assert that all changes adhere to the asked for style
-    exec git diff --exit-code
+    exec git diff --exit-code -- . ":!3rdparty"
```
This change doesn't belong to b22cbf71a77625200d543c21a99c9d791e6e8bc8 (_Recalculate position so it is based on the index of unicode codepoints_). It should be in a commit of its own, or at least in the commit that updates `json-spirit`.
